### DISCUSSION TITLE
How to delete a prop from node ?

### DISCUSSION THREAD

**Discussion by ProPhycient5** [prevwong/craft.js#388] [Q&A] [answered]
Taking below code as example to put my query in better way:
```
import {useEditor} from "@craftjs/core";

const Example = () => {
  const { selectedNodeId, actions: {setProp} } = useEditor((state) => ({
    selectedNodeId: state.events.selected
  }));

  return (
    <a
      onClick={_ => {
        setProp(selectedNodeId, props => {
          props.text = "new value";
        });
      }}
    >
      Update
    </a>
  )
}
```
Here, we see that `text` is already exist as one of the prop or even we can set new prop using `actions: {setProp}` using `useNode` or  `useEditor`. My question is : How can we delete `text` permanently from that node ?? 
I don't want to keep by assigning its value as `undefined` or `empty`.

Please do comment below 👇🏼👇🏼👇🏼👇🏼👇🏼 your solution taking above code as example

**Comment by prevwong** [comment, marked as answer]
```tsx
actions.setProp(prop => {
  delete prop.text;
});
```

**Reply by ProPhycient5** [reply]
Thanks @prevwong for your kind help 🙏🏼

**Reply by ProPhycient5** [reply]
I found one tip where, it is recommend to use spread operator instead of `delete` keyword in order to remove property from object.

Have a look 👇🏼 and confirm me whether I should go above code or any change is required, to make bug-proof.
![prev-1](https://user-images.githubusercontent.com/71059909/157677901-1dde7ce2-5090-4690-9bea-25779b205e24.JPG)

**Reply by ankri** [reply]
tl;dr: IMO these rules don't apply because Craft.js uses immer. Which takes care of the immutability.

I imagine SimonHoiberg's tip is about immutability. Because you would mutate the object `simon` by deleting the property. By using the rest operator you would create a new object `newSimon` without mutating `simon`. I really recommend following this suggestion and trying to write as much immutable code as possible, since it's easier to reason with. 

On the other hand `actions.setProp` uses immer, which takes care of the immutability. You can write "mutable" code but immer's magic will take care of the immutability. Have a look at the official [documentation](https://immerjs.github.io/immer/update-patterns/). They also recommend using `delete` there :)

**Reply by ProPhycient5** [reply]
Thank @ankri for the confirmation and clarification. Glad to learn new thing about immutability.
